add custom leo image tab [SATURN-1090]

zarsky-broad opened the pull request:
https://broadworkbench.atlassian.net/browse/SATURN-1090

Ready for review

zarsky-broad committed:
select a leo image

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

zarsky-broad committed:
break up long line

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

zarsky-broad committed:
PR comments

zarsky-broad committed:
multiple languages of packages

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

zarsky-broad committed:
one more deps file, wording with Adrian

zarsky-broad committed:
better paren placement

ehanna4 committed:
Add tab for custom env

Add tab for custom env
- still todo many things but visual is mostly there now

zarsky-broad committed:
Merge branch 'dev' into eh-add-custom-env

zarsky-broad committed:
spurious newlines

zarsky-broad committed:
everything but the link targets

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

zarsky-broad committed:
cleanup

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

ehanna4 committed:
Update links

ehanna4 committed:
Fix lint issues

zarsky-broad committed:
add image validation regexp

zarsky-broad committed:
cleanup

zarsky-broad committed:
linebreaks

zarsky-broad committed:
add validation

zarsky-broad committed:
alignment, performance

ehanna4 committed:
Prevent elements shifting

- Contain element-moving-when-error-mssg-shows issue to top grid section

ehanna4 committed:
Remove element-shift when error msg appears

ehanna4 committed:
Split out NewClusterModal to own file

panentheos commented on src/data/leo-images.js:
```diff
@@ -22,4 +22,16 @@ const leoImages = [
+
+// see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp#Description
+// for why it's preferable to create a regex from a regexp here, rather than from a string
+export const imageValidationRegexp = new RegExp(new RegExp(referenceRegexp))
```
This regexp seems fairly permissive, e.g. the string `aaa` is valid. Is that intended?

ehanna4 replied:
Hi, yes this is the intention. This regex comes from docker's github for checking a docker image. The backend will also be doing some validation but here we are just looking to check it's in the correct format for an image

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -0,0 +1,431 @@
+      profile: matchingProfile ? matchingProfile.name : 'custom',
+      jupyterUserScriptUri: '',
+      selectedLeoImage: leoImages[0].image,
+      isCustomEnv: false, customEnvImage: '', showingCustomWarning: false,
```
since `viewingPackages` and `showingCustomWarning` are mutually exclusive, we could represent them with a single `mode` state that represents which view we're in.

ehanna4 replied:
Addressed in latest commit, thanks!

ehanna4 committed:
Add link for safe image creation, change image path name and placeholder example

ehanna4 committed:
Update TitleBar component and link change

ehanna4 committed:
update link

panentheos commented on src/components/TitleBar.js:
```diff
@@ -10,13 +10,15 @@ const TitleBar = ({ onPrevious, title, onDismiss, titleExtras }) => {
-    }, [icon('arrowLeft', { size: 17 })]),
-    div({ style: { fontSize: '0.875rem', fontWeight: 600, marginLeft: onPrevious ? 'auto' : undefined } }, [title]),
+    }, [icon('arrowLeft', { size: 20 })]),
+    (onPrevious && onDismiss) ? div({ style: { fontSize: '0.875rem', fontWeight: 600, marginLeft: onPrevious ? 'auto' : undefined } }, [title]) : undefined,
```
This means the buttons will swap sides depending on the situation? That seems like it could be a confusing UX. What was the goal here?

ehanna4 replied:
The goal was that if there was only going to be only one action available (either an 'x' exit or '<-' back) that this button would always be on the right side. This came from the updated mock up and a discussion with Lou, with the goal of a more consistent look to the UI

ehanna4 replied:
mock up with updated look: https://projects.invisionapp.com/d/main?origin=v7#/console/16427254/383729206/preview

ehanna4 replied:
Flipped it back after consultation with @zarsky-broad

ehanna4 committed:
Update link

ehanna4 committed:
Do switchcase instead of cond for drawer views

- Changed to only look at one piece of state for altering the drawer views rather than two

ehanna4 committed:
Flipping title bar back

Co-Authored-By: Isaac Zarsky <zarsky-broad@users.noreply.github.com>

panentheos reviewed (approved):
Looks good, thanks!

ehanna4 committed:
Merge branch 'dev' into eh-add-custom-env

ehanna4 committed:
Merged in changes from dev and fixed link that changed

ehanna4 committed:
Bring over change that got lost in dev merge

ehanna4 committed:
Merge branch 'dev' into eh-add-custom-env

ehanna4 committed:
Fix performance issue of regex

ehanna4 committed:
Add comment about choice to use different regex piece

zarsky-broad committed:
waaaaaay simpler regex

Co-Authored-By: ehanna4 <ehanna4@users.noreply.github.com>

panentheos commented on src/components/NewClusterModal.js:
```diff
@@ -0,0 +1,432 @@
+            ' to select another image.'
+          ]),
+          div({ style: { display: 'flex', justifyContent: 'flex-end' } }, [
+            h(ButtonSecondary, { style: { marginRight: '2rem' }, onClick: () => this.setState({ viewMode: 'undefined' }) }, ['Back']),
```
```suggestion
            h(ButtonSecondary, { style: { marginRight: '2rem' }, onClick: () => this.setState({ viewMode: undefined }) }, ['Back']),
```

panentheos reviewed (approved):
Looks good, see comment.

ehanna4 committed:
Update src/components/NewClusterModal.js

Co-Authored-By: Brett Heath-Wlaz <bheathwl@broadinstitute.org>
